dio/lib/src/dio_event_processor.dart

```diff
@@ -0,0 +1,110 @@
+// ignore_for_file: implementation_imports, public_member_api_docs
+
+import 'dart:async';
+
+import 'package:dio/dio.dart';
+import 'package:sentry/sentry.dart';
+import 'package:sentry/src/sentry_exception_factory.dart';
+import 'package:sentry/src/sentry_stack_trace_factory.dart';
+
+class DioEventProcessor implements EventProcessor {
+  DioEventProcessor(this._options, this._maxRequestBodySize);
+
+  final SentryOptions _options;
+  final MaxRequestBodySize _maxRequestBodySize;
+  late final sentryExceptionFactory = SentryExceptionFactory(
+    _options,
+    SentryStackTraceFactory(_options),
+  );
+
+  @override
+  FutureOr<SentryEvent?> apply(SentryEvent event, {dynamic hint}) {
+    final dynamic dioError = event.throwable;
+    if (dioError is! DioError) {
+      return event;
+    }
+
+    try {
+      final mechanism = Mechanism(
+        type: 'DioIntegration',
+        description: 'This is the inner exception of the DioError',
+      );
+      final throwableMechanism = ThrowableMechanism(mechanism, dioError.error);
```

Comment:
What if we always use the `dioError.error` for the `dio` integration, isn't that any better? is there info on `dioError` that's not on `dioError.error`?

Reply:
Yep, there's info on `DioError` which is not available on `DioError.error`. Also, `DioError.error` is not always available.

Reply:
Makes sense to report both whenever possible then.

I'm wondering if we can make it without the EventProcessor.
What if we pass a List of exceptions to `event.throwable` since its a `dynamic` field and in the `sentryExceptionFactory` we special case if it's a `List`, we traverse the list converting to `SentryException`, didn't try, but it could work.

The downside is if somebody does `throw randomList` and this List isn't this special list of exceptions, but rather an App's list, we need to consider that too.

Reply:
We don't have the `DioError` available earlier. The `SentryDioClientAdapter` is not aware of these errors. The `DioError`s usually are manually caught and captured or come from the zones uncaught handler.

Reply:
But then we would have to know how to create list from `DioError` inside the `sentry` sdk. But that doesn't know about Dio.

Is there a way to register a `SentryExceptionFactory` for specific exceptions?

Reply:
I see, then an event processor makes sense, we just need to add it automatically, so fewer steps for the final user.
A thing to consider is, an App may have more than one `dio` instance, so when `addSentry` is called, we should not register the event processor twice.

Reply:
> I see, then an event processor makes sense, we just need to add it automatically, so fewer steps for the final user.

See https://github.com/getsentry/sentry-dart/pull/718#discussion_r792821809

> A thing to consider is, an App may have more than one dio instance, so when addSentry is called, we should not register the event processor twice.

Good point.